### ISSUE TITLE
Single-file kernel load from ecm repo?

### ISSUE THREAD

**Issue by ecm-pushbx** [SvarDOS/edrdos#28]
Hi, I noticed your recent commits don't include the patches to use single-file load via lDOS iniload / drload + drkernpl. In case you missed it, I commented this some on https://pushbx.org/ecm/dokuwiki/blog/pushbx/2024/0107_enhanced_dr-dos_single-file_load

Did you intentionally not include these patches? Do you have additional questions or is there something else I can do to help you?

**Comment by boeckmann** [comment]
Hi! I intentionally delayed the inclusion of the single-file patches, because

1. I wanted to release at least one more-or-less working dual-file package for SvarDOS before importing it, and
2. wanted to play around with lDebug boot capabilities before importing it (I am currently studying the extensive lDebug manual :-) )

When I import the changes for single-file load, do I have to provide a new SYS command with the SvarDOS package? I also noticed you opened a ticket regarding the boot protocol at https://github.com/FDOS/kernel/issues/119. Next, I am not sure if I should LZMA compress the kernel for SvarDOS, because you mentioned some 30 second decompression time. But I have to test this and weight the options against each other. If I use it uncompressed, I should at least test if this zero-compression would decrease the file size, so implementing an open source version of compbios/compbdos, probably as a single tool, would be high on the priority list. All in all I have to invest some time to understand how this all fits together.

I am also somewhat distracted by doing some work on command.com, fixing annoyances as I encounter them. But I think I am nearly through with this.

**Comment by ecm-pushbx** [comment]
> Hi! I intentionally delayed the inclusion of the single-file patches, because
> 
> 1. I wanted to release at least one more-or-less working dual-file package for SvarDOS before importing it, and

Fair.
 
> 2. wanted to play around with lDebug boot capabilities before importing it (I am currently studying the extensive lDebug manual :-) )

If you come across any questions not answered in the manual feel free to drop me a note ^^ (Here or in a mail or on the forum or the blog.)

> When I import the changes for single-file load, do I have to provide a new SYS command with the SvarDOS package?

All EDR-DOS / FreeDOS SYS programs should work if you have the user pass `/bootonly /k edrdos.com` or one of the other filenames. It's your call if you want to build and host a new revision of FreeDOS SYS with a load protocol that defaults to one of the new names. (The current FreeDOS kernel sources contain *most* of the EDR-DOS patches in `#ifdef` conditionals I believe.)

> I also noticed you opened a ticket regarding the boot protocol at [FDOS/kernel#119](https://github.com/FDOS/kernel/issues/119).

Yes, but as mentioned this does not affect the drload or iniload stages as used by my EDR-DOS and lDebug. I will likely submit a patch to FreeDOS to just have the FAT32 loaders always pass both registers at a later time. This would also make it so you could compile EDR-DOS SYS from the shared sources with the existing `#ifdef` blocks without needing to patch the boot loaders to match the original EDR SYS's, yet work just like that to pass DL too.

> Next, I am not sure if I should LZMA compress the kernel for SvarDOS, because you mentioned some 30 second decompression time. But I have to test this and weight the options against each other.

In that blog post I actually mentioned it can take several minutes, here:

> However, depacking on a low-end machine (eg NEC V20) may take several minutes to complete.

This is based on the performance of lDebug release 5 tested on an 1 MiB model of the HP 95LX, powered by a "5.37 MHz NEC V20". (I don't think you can boot another DOS easily on that particular device, but it does provide some reference for low-end machines.) [In the blog post comparing the depack speed](https://pushbx.org/ecm/dokuwiki/blog/pushbx/2023/0321_cpu_performance_comparison#the_hp_95lx) I found that the LZMA-302eos depacker needed more than two minutes. This is with a depacked kernel stage of size 106 KiB, compressed to an image of 62 KiB for LZMA-302eos (not including iniload or depacker), and an image of 69 KiB for LZSA2 which is also mentioned in the performance comparison.

(If you want to change the example of my EDR-DOS mak.sh script to use LZSA2 compression instead, you would have to [pass -D_LZSA2 to inicomp.asm instead](https://hg.pushbx.org/ecm/ldebug/file/024ec1c080a2/source/mak.sh#l508) and [use the lzsa command line tool to compress](https://hg.pushbx.org/ecm/ldebug/file/024ec1c080a2/source/mak.sh#l571), which is found [in this repo](https://github.com/emmanuel-marty/lzsa). The lDebug mak.sh script is the canonical source of truth about inicomp use.)

I did improve things for the LZMA-302eos depacker a bit [by enabling its _COUNTER define](https://hg.pushbx.org/ecm/edrdos/rev/91f95e817381), that is, having it display a progress indicator (by default a dot is displayed [at every 128th step](https://hg.pushbx.org/ecm/inicomp/file/9725324008b6/lzd.asm#l25)). About four lines of 80 columns each are filled by the 2024 January revision when the depacker runs to completion.

> If I use it uncompressed, I should at least test if this zero-compression would decrease the file size, so implementing an open source version of compbios/compbdos, probably as a single tool, would be high on the priority list.

It likely would decrease the size. However, you would have to build the (BIO module) kernel twice if you want to use it with and without the compression to build different versions of the kernel file.

> All in all I have to invest some time to understand how this all fits together.

Fair enough, just wanted to check in.

> I am also somewhat distracted by doing some work on command.com, fixing annoyances as I encounter them. But I think I am nearly through with this.

Something I would like to see in the shell is `SET /E variable=command` support modelled on FreeCOM's. Basically it internally runs the command, redirecting its stdout like `command > file` would, then grabbing the first line from the temporary file to use as the content of the variable. I would probably be able to do it on my own but I would also welcome it if you worked on this.

**Comment by ecm-pushbx** [comment]
SET /E issue created in response to my comment: https://github.com/SvarDOS/edrdos/issues/30

**Comment by ecm-pushbx** [comment]
Can you show the exact commands and files that you used? Do your sizes include depackers for these formats?

**Comment by boeckmann** [comment]
The exoraw command was (DRBIO.SYS):
```
exoraw DRBIO.SYS -o DRBIO.TST
```

Heatshrink:
```
heatshrink -w 8 DRBIO.SYS DRBIO.TXT
```
The depacker size is not included. This was a first test to see how compression generally performs.

**Comment by boeckmann** [comment]
What is interesting is that exoraw performs much better on actual X86 code. While heatshrink does a good job of getting rid of the remaining zero-filled areas in DRBIO.SYS, compression of DRDOS.SYS is quite "bad".

**Comment by boeckmann** [comment]
The exomizer version I use is from commit https://bitbucket.org/magli143/exomizer/commits/df77c879ce2addc027043e5e80e3992a2ec99eb9. Version 3.1.1.

**Comment by ecm-pushbx** [comment]
I do have support for both exomizer and heatshrink in inicomp already. Also apl/apultra and of course the lzip I already used in my EDR-DOS kernel build. I'm using exomizer-3.0.2.zip, also from that bitbucket repo. Here's some numbers gathered from building [the current tip of lDebug](https://hg.pushbx.org/ecm/ldebug/rev/78e49c57a05f), the non-debuggable non-DPMI triple-mode executable ldebug.com: (The sizes are for the entire executable, including lDOS boot iniload and the inicomp depacker stage.)

```
ldebug/source$ INICOMP_SPEED_TEST=128 use_build_decomp_test=1 INICOMP_METHOD="lzd exodecr apl heatshrink lzsa2" ./mak.sh
[...]
ldebug/source$ LC_ALL=C sort ../tmp/debug.siz
   79872 bytes ( 60.70%), method              lzd
   83968 bytes ( 63.81%), method              apl
   84480 bytes ( 64.20%), method          exodecr
   86528 bytes ( 65.75%), method            lzsa2
   99840 bytes ( 75.87%), method       heatshrink
  131584 bytes (100.00%), method             none
ldebug/source$ LC_ALL=C sort ../tmp/debug.spd
    1.46s for 128 runs (   11ms / run), method            lzsa2
    3.00s for 128 runs (   23ms / run), method              apl
    3.84s for 128 runs (   30ms / run), method          exodecr
    5.00s for 128 runs (   39ms / run), method       heatshrink
    9.56s for 128 runs (   74ms / run), method              lzd
ldebug/source$ hg id
78e49c57a05f tip
```

The speed test uses dosemu2 with KVM on an amd64 Debian system. Here's [the depack stage size displayed for each](https://hg.pushbx.org/ecm/inicomp/file/ebc8a12f4841/inicomp.asm#l1334) of those:

```
../../inicomp/lzd.asm:698: warning: localvariables has 14688 bytes [-w+user]
../../inicomp/inicomp.asm:1323: warning: inilz: 3024 bytes used for depacker [-w+user]
../../inicomp/inicomp.asm:1323: warning: iniexo: 1472 bytes used for depacker [-w+user]
../../inicomp/inicomp.asm:1323: warning: iniapl: 1344 bytes used for depacker [-w+user]
../../inicomp/inicomp.asm:1323: warning: inihs: 1360 bytes used for depacker [-w+user]
../../inicomp/inicomp.asm:1323: warning: inilzsa2: 1424 bytes used for depacker [-w+user]
```

And here's the compression commands used for all of the supported formats: https://hg.pushbx.org/ecm/ldebug/file/78e49c57a05f/source/mak.sh#l530

**Comment by ecm-pushbx** [comment]
By the way, the best heatshrink option turns out to be `-w 14 -l 4` for this lDebug build. The mak.sh script tests all valid combinations of the parameters.

**Comment by boeckmann** [comment]
Yes, 32 corresponds to bit 5. No idea why I wrote 8 :-/

**Comment by mateuszviste** [comment]
While these heatshrink / exoraw options are very interesting, could you tell what is the reason that the DR kernel cannot be UPXed like how the FreeDOS folks do? UPX is quite efficient when it comes to compressing x86 code. I assume there is an obvious technical limitation but I'm just to dumb to understand it on my own, so please consider this to be a "please educate me" question :)

**Comment by ecm-pushbx** [comment]
I have worked a little with the FD kernel UPX support, for example [creating an offline depacker](https://github.com/FDOS/kernel/issues/75) and [changing the format and placement of the CONFIG block and stubs](https://github.com/FDOS/kernel/pull/78).

However, I prefer working with my own depackers and inicomp stage, which offer more choices than UPX and also work with triple-mode executables (kernel, DOS device driver, DOS application). (Executables larger than 64 KiB are supported with the new DOS/EXE exeflat format from that PR of mine that was merged.) The usage conditions and licenses are clearer as well. (I have a reminder somewhere to complain on the freedos-devel list that the DOS UPX releases mirrored by FD use the nonfree NRV compression library. Not a big priority but not nice either.) (The usage conditions are a reason for my offline depacker as UPX states you should be able to unpack the executable, and not to modify their depackers.)

**Comment by ecm-pushbx** [comment]
Reviewing the new exeflat format, it also seems to depend on the kernel being loaded at segment 60h which is another hurdle. Probably not that difficult to change but I'd have to study it more.

**Comment by ecm-pushbx** [comment]
> but I understand that whether it is UPX or a custom packer, in both cases some serious work must happen first,

Well, for lDOS inicomp + lDOS drload/iniload very little work is needed, just pick my single-file kernel patches and (if other format than lzip desired) adapt the compression commands from lDebug's main mak.sh.

**Comment by boeckmann** [comment]
As a sidenote: I wanted to test how well exomizer compresses the FDISK executable in comparison to UPX, and here are the results:

exeraw: 38321 bytes, 71994 uncompressed
UPX: 39247 bytes, but this includes the decompression stub

so all-in-all on par, I would say. Within ~500 bytes difference considering decompressor size to be added for exeraw.

**Comment by ecm-pushbx** [comment]
I added support for Exomizer's new -P 32 flag in https://hg.pushbx.org/ecm/inicomp/rev/98918a411ab5

I also added some auto-detection to pass the correct `-D_P=` switch to NASM, instead of defaulting to 7 when no `INICOMP_EXOMIZER_P` variable is given: https://hg.pushbx.org/ecm/ldebug/rev/bcaea57de05c

Doubling the -P switch is needed to work around a bug in older Exomizer versions: https://hg.pushbx.org/ecm/ldebug/rev/3786ef6709ef

**Comment by boeckmann** [comment]
I yesterday had a look at the exomizer source code. While the decoding algorithm is fairly straightforward, I am struggling to reverse engineer how the compressor actually works. I have not found an algorithmic description, and the source is basically without comments. Thats a little sad, because I really would like to know how this thing works.

Apart from that, digging a little deeper into this kernel-compression topic I am starting to wonder if using an alternative algorithm like LZSA2 would be a better fit for the single-file compressed version of the EDR kernel, because according to https://github.com/emmanuel-marty/lzsa, decompression should be significally faster. And ECMs measurements above indicate a decompression speedup of 2.5 times for LZSA2 in comparison to exomizer (though it is not measured on an 8088), buyed by moderate increase in file size.

**Comment by ecm-pushbx** [comment]
https://pushbx.org/ecm/dokuwiki/blog/pushbx/2024/0319_early_mid_march_work

> Further, I have considered a new repo which will be based on the mak script of lDebug and will provide the scripting needed to wrap and optionally compress the Enhanced DR-DOS single-file kernel or the FreeDOS kernel. This will also eventually lead to CONFIG block support in inicomp and fdkernpl (part of ldosboot).

I didn't get around to this yet. The lDebug mak script allows to build several compression formats in the same run, so users could choose either apl, Exomizer, LZMA-lzip, or faster formats like LZSA2.

**Comment by ecm-pushbx** [comment]
Current EDR-DOS builds from me use [the new kernwrap script](https://hg.pushbx.org/ecm/kernwrap/) which is derived from lDebug's mak script. So in the tmp/ subdirectories users [may choose alternatives](https://hg.pushbx.org/ecm/edrdos/file/729bae2f79d7/cfg.sh#l26) of the compressed edrpack.sys (drload) or edrpack.com (iniload) files, including the very fast LZSA2, or LZ4, or heatshrink, or the also strong apl or exomizer.

**Comment by ecm-pushbx** [comment]
> I yesterday had a look at the exomizer source code. While the decoding algorithm is fairly straightforward, I am struggling to reverse engineer how the compressor actually works. I have not found an algorithmic description, and the source is basically without comments. Thats a little sad, because I really would like to know how this thing works.

I don't know much about any of the compression methods I use, they all just had usable enough depacker code to depack for me to port it or detailed descriptions of the resulting format so that I could implement a depacker on my own.

This difference is what determines my licensing [as stated in my lDebug manual](https://pushbx.org/ecm/doc/ldebug.htm#additionalusageconditions). So the depackers for LZ4, Snappy, Heatshrink, and LZO were written by me. The remaining ones I ported, that is BriefLZ, Exomizer, X, Lzd (lzip), LZSA2, apl, bzpack.

I haven't had any good ideas for compressing data on my own. I did experiment some with "bit runs" which was basically a way to record how many consecutive bits of the same value occur. This turned out to be a great way to *expand* the size of a file.

The other idea I've had is to use genetic programming to obtain an algorithm/program that reconstructs the original data. Also null useful results so far.

At one point I tried out a neural net based compression method someone else wrote (unfortunately don't recall the name) but this had high memory consumption and was very slow. And it was symmetric in time and space, ie the depacker was as slow as the packer.

**Comment by boeckmann** [comment]
@ecm-pushbx  I will try to create a new EDR package for SvarDOS in the next days, using your current single-file kernel. If I understand correct, the edrdos.sys and edrpack.sys would work with the FreeDOS SYS provided loader in DRDOS boot protocol mode? Would it be much effort for you to provide me with a LZSA2 version of the single file? Otherwise I would have to setup a Linux VM with an appropriate build environment (DOSEMU etc.).

**Comment by ecm-pushbx** [comment]
> @ecm-pushbx I will try to create a new EDR package for SvarDOS in the next days, using your current single-file kernel. If I understand correct, the edrdos.sys and edrpack.sys would work with the FreeDOS SYS provided loader in DRDOS boot protocol mode?

All files (edrdos/edrpack).(sys/com) work with FreeDOS loaders as replacement for kernel.sys or the original EDR-DOS loaders (based on FreeDOS's) as replacement for drbio.sys. The .com files also work as MS-DOS, IBM-DOS, Multiboot, lDOS / RxDOS kernels. To your specific question, yes, both of the .sys files will work with EDR SYS's default loader. (But SYS may expect the drdos.sys file to exist or be copied over as well. This is not needed for us.)

(The exception is that if you append arbitrary data such as a .zip file to the kernel then it will at some point (eg > 128 KiB) fail to work as FreeDOS/EDR-DOS load. So you shouldn't ever do that to the .sys files as they will be useless then. The .com files can still be loaded in one of the other load protocols even with data > 1 MiB.)

> Would it be much effort for you to provide me with a LZSA2 version of the single file? Otherwise I would have to setup a Linux VM with an appropriate build environment (DOSEMU etc.).

It is already shipped in https://pushbx.org/ecm/download/edrdos.zip in the subdirectory `tmp/sa2/` which has an edrpack.sys and edrpack.com

**Comment by ecm-pushbx** [comment]
Oops, forgot to submit the prior reply until now.

**Comment by boeckmann** [comment]
Cool, thanks. Actually I already downloaded the archive but did not expect the LZSA2 kernel in the tmp folder :)

**Comment by boeckmann** [comment]
I added a `/OEM:EDRPACK` option to FreeDOS SYS, which makes the loader search for `EDRPACK.SYS`. @ecm-pushbx is EDRPACK.SYS the final name for it? Kernel loads fine. I might create a pull request at FreeDOS for the addition.

I will remove the DRSYS.COM file from the SvarDOS kernel package, and instead would replace the SvarDOS SYS command with this one. Kernel then has to be installed via `SYS /UPDATE /OEM:EDRPACK` or `SYS /UPDATE /OEM:EDR` (for the dual-file version), with the default installing the FreeDOS kernel. But this would be ok I think. Should not overburden the ordinary user able to install SvarDOS :)

**Comment by mateuszviste** [comment]
This sounds awesome, thank you Bernd & ECM!

About the "SYS /UPDATE /OEM:xxxx" business - wouldn't it be possible for SYS to figure out which kernel it installs? For example:
If C:\KERNEL.SYS is present -> FreeDOS mode
If C:\EDRPACK.SYS is present -> EDR-DOS packed
If C:\DRBIO.SYS etc

Of course it won't work if someone has different variants present in C:\, so an /OEM way to enforce a specific kernel would still be required, but such automatic detection would at least cover the most obvious scenario of a single-kernel system.

**Comment by boeckmann** [comment]
> About the "SYS /UPDATE /OEM:xxxx" business - wouldn't it be possible for SYS to figure out which kernel it installs? For example: If C:\KERNEL.SYS is present -> FreeDOS mode If C:\EDRPACK.SYS is present -> EDR-DOS packed If C:\DRBIO.SYS etc
> 
> Of course it won't work if someone has different variants present in C:, so an /OEM way to enforce a specific kernel would still be required, but such automatic detection would at least cover the most obvious scenario of a single-kernel system.

Well this is another option that is supported "out-of-the-box". If you SYS from the directory that only contains the EDR kernel (the package directory) it will install it without the user being forced to use /OEM. Otherwise the search order is:
- FreeDOS kernel.sys
- Single-file EDRPACK.SYS
- "ordinary" DRBIO.SYS / DRDOS.SYS combination

**Comment by ecm-pushbx** [comment]
> I added a `/OEM:EDRPACK` option to FreeDOS SYS, which makes the loader search for `EDRPACK.SYS`. @ecm-pushbx is EDRPACK.SYS the final name for it? Kernel loads fine. I might create a pull request at FreeDOS for the addition.
> 
> I will remove the DRSYS.COM file from the SvarDOS kernel package, and instead would replace the SvarDOS SYS command with this one. Kernel then has to be installed via `SYS /UPDATE /OEM:EDRPACK` or `SYS /UPDATE /OEM:EDR` (for the dual-file version), with the default installing the FreeDOS kernel. But this would be ok I think. Should not overburden the ordinary user able to install SvarDOS :)

Actually I would like it if loading / using edrdos.com was also supported. But `/OEM:EDR` is taken and adding a distinct `/OEM:EDRDOS` would be confusing. Therefore I would suggest `/OEM:LEDR` for loading lDOS iniload edrdos.com and probably `/OEM:LEDRPACK` for loading lDOS drload edrpack.sys. This could be symmetrical to a to-be-added `/OEM:LMS` and `/OEM:LMSPACK` (as my MS-DOS v4 fork files are called lmsdos.com and lmspack.sys).

Further, I would prefer to use either segment 60h (like FreeDOS load) or segment 200h by default for LEDR, LEDRPACK, LMS, or LMSPACK. This avoids possible problems with DMA that has to cross a 64 KiB boundary. (The FreeDOS original FAT12 and FAT16 loaders do their sector reads into a temporary buffer, but lDOS boot's loaders in FreeDOS compatible modes do not. Segment 60h implies that a 512-byte sector read is never crossing a 64 KiB boundary, while with segment 70h it will if the file is about 64 KiB in size. Segment 200h similarly avoids the problem.)

**Comment by boeckmann** [comment]
My current plan on the single-file kernel is as follows: I changed DRBIO so that the zero-uncompression code is placed in the deblocking-buffer near the start at DRBIO instead of the init code far into the file. This makes it possible to compress nearly the whole DRBIO, and brought down DRBIO size from ~36K to ~24K.

Next I want to produce a combined kernel file, with DRBIO and DRSYS combined and zero compressed. Sadly, this brings us only down to ~60K, but it is a start.

When I got this to work, we can replace the zero compression with any compression algorithm which decompressor does not require more than 512 bytes to still fit into the deblocking buffer.

The benefit of the whole procedure is that this would not depend on external components like a separate uncompressor stage prepended to the kernel.

There seem to be plenty or ready-to-go uncompressor implementations that fit into the 512 byte target. Like @ecm-pushbx pointed out, LZMA is way too slow for ancient machines. We should choose another one. For simplicity, I opt for sticking with a single algorithm, and let the uncompressor code in place for the compressed and uncompressed kernel (the deblocking space is there anyway). If the user wants to experiment with different compressors he/she can still use the ecm implementation.

**Comment by ecm-pushbx** [comment]
> Next I want to produce a combined kernel file, with DRBIO and DRSYS combined and zero compressed

Think you meant DRDOS here.

**Comment by boeckmann** [comment]
> Think you meant DRDOS here.

Yes.

**Comment by ecm-pushbx** [comment]
> Like @ecm-pushbx pointed out, LZMA is way too slow for ancient machines.

Relevant reference: https://github.com/SvarDOS/edrdos/issues/69#issuecomment-2244834392

**Comment by ecm-pushbx** [comment]
Presumably you want to retain loading only with the EDR-DOS load protocol? As that implies the entire file is loaded so you don't need an initial loader or lDOS iniload stage (akin to Microsoft MSLOAD). Will you also support loading at better segments than 70h like 60h (FreeDOS load) and 200h?

**Comment by boeckmann** [comment]
> Presumably you want to retain loading only with the EDR-DOS load protocol? As that implies the entire file is loaded so you don't need an initial loader or lDOS iniload stage (akin to Microsoft MSLOAD).

Yes, that is the plan. Or switching to the FreeDOS load protocol.

> Will you also support loading at better segments than 70h like 60h (FreeDOS load) and 200h?

I am not yet sure about the consequences loading it to 60h. But I think it would be desireable to eventually utilize the same load protocol as the FreeDOS kernel, so switching the kernel would merely mean replacing the kernel binaries.

**Comment by mateuszviste** [comment]
> I am not yet sure about the consequences loading it to 60h. But I think it would be desireable to eventually utilize the same load protocol as the FreeDOS kernel, so switching the kernel would merely mean replacing the kernel binaries.

That would be indeed very good, ie. not having to do a SYS after replacing kernel.sys by edrpack.sys (or maybe could it even be called kernel.sys as well?) It would definitely make it simple for SvarDOS user to switch between kernels.

BTW may I ask why a 60h is a better segment than 70h or 200h? Is it a matter of limiting memory fragmentation?

**Comment by ecm-pushbx** [comment]
> BTW may I ask why a 60h is a better segment than 70h or 200h? Is it a matter of limiting memory fragmentation?

60h and 200h are both better than 70h. That's because they are aligned in memory at the normal sector size (512 Bytes) or for 200h even up to the maximum sector size that a DOS can generally handle (8 KiB). This is desirable because ISA DMA may be unable to read a sector so that it crosses a 64 KiB boundary in memory (ie the boundary at 10000h, 20000h, etc).

[Quoth me](https://github.com/SvarDOS/edrdos/issues/28#issuecomment-2118901558):

> Further, I would prefer to use either segment 60h (like FreeDOS load) or segment 200h by default for LEDR, LEDRPACK, LMS, or LMSPACK. This avoids possible problems with DMA that has to cross a 64 KiB boundary. (The FreeDOS original FAT12 and FAT16 loaders do their sector reads into a temporary buffer, but lDOS boot's loaders in FreeDOS compatible modes do not. Segment 60h implies that a 512-byte sector read is never crossing a 64 KiB boundary, while with segment 70h it will if the file is about 64 KiB in size. Segment 200h similarly avoids the problem.)

00600h / segment 60h / entrypoint 60h:0 is also the standard for FreeDOS's kernel.sys, albeit FreeDOS's SYS provides [the /L switch](https://github.com/FDOS/kernel/blob/db43a4b3438047caccec220e349ede57fb24f88f/sys/sys.c#L528) to modify the default loader to use a different segment (still with entrypoint at +0:0 where the "plus" means displacement from the load address segment).

200h is a valid choice for SYS's /L switch as well. Other than that, it matches the new default for the lDOS load protocol but that actually loads to 02000h / segment 200h / entrypoint 200h:400h so its entrypoint doesn't match the FreeDOS style one. (And the lDOS entrypoint has [some more requirements](https://pushbx.org/ecm/doc/ldosboot.htm#protocol-sector-iniload).)

**Comment by boeckmann** [comment]
Ok, I have three of four cases working: uncompressed and compressed dual-file, and uncompressed single-file.

The zero-compressed single-file is causing some trouble at the moment. In the next days I will verify if the compressed single-file is valid by uncompressing it and comparing with the uncompressed single-file. If the files match it has to be some problem with the kernel uncompressor.

Though it seems the kernel decompresses at least parts of the single-file kernel good. If I change offset 5 of the binary from 0x81 to 0x01 (kernel flag changed from compressed single-file to compressed dual-file), I get a working DOS prompt. It then reads DRDOS.SYS from disk despite the kernel being single-file because the flag instructs it to do so...

**Comment by mateuszviste** [comment]
> 60h and 200h are both better than 70h. That's because they are aligned in memory at the normal sector size (512 Bytes) or for 200h even up to the maximum sector size that a DOS can generally handle (8 KiB). This is desirable because ISA DMA may be unable to read a sector so that it crosses a 64 KiB boundary in memory (ie the boundary at 10000h, 20000h, etc).

Thank you for your thorough explanations. I needed some time to process it. :-)

So I understand this is all about the boot loader being able to perform DMA copies when loading the kernel file from disk into memory. Such load operation being carried sector-after-sector, and the (logical, or "virtual") sector size being realistically always 512 bytes, loading the kernel file at seg 60h makes no 512-bytes chunk cross the 64K boundary, while loading at seg 70h would make the 124th sector land at address 0xFF00, thus crossing the 64K boundary, possibly failing due to legacy DMA limitations. All this does not matter of course as long as the kernel file (compressed, ie. as seen by the boot loader) is less than 63744 bytes in size. Did I get it right?

**Comment by ecm-pushbx** [comment]
> > 60h and 200h are both better than 70h. That's because they are aligned in memory at the normal sector size (512 Bytes) or for 200h even up to the maximum sector size that a DOS can generally handle (8 KiB). This is desirable because ISA DMA may be unable to read a sector so that it crosses a 64 KiB boundary in memory (ie the boundary at 10000h, 20000h, etc).
> 
> Thank you for your thorough explanations. I needed some time to process it. :-)
> 
> So I understand this is all about the boot loader being able to perform DMA copies when loading the kernel file from disk into memory.

Yes. And ISA DMA particularly may be used to access diskette drives.

> Such load operation being carried sector-after-sector, and the (logical, or "virtual") sector size being realistically always 512 bytes, loading the kernel file at seg 60h makes no 512-bytes chunk cross the 64K boundary, while loading at seg 70h would make the 124th sector land at address 0xFF00, thus crossing the 64K boundary, possibly failing due to legacy DMA limitations.

Yes, exactingly.


> All this does not matter of course as long as the kernel file (compressed, ie. as seen by the boot loader) is less than 63744 bytes in size. Did I get it right?

I didn't check the number but this is broadly correct. However, if you have a cluster size of say 8 KiB then the file size as experienced by the boot sector loader is always rounded up to a full cluster, so loading to 00700h won't cross the 10000h boundary only if the file is at most 56 KiB sized.

Other than that I always try supporting large files; the FreeDOS load protocol is now good for files up to 128 KiB (with cluster size up to 128 KiB too) or for smaller clusters (eg 2 KiB) up to 134 KiB.

**Comment by boeckmann** [comment]
Zero-compressed single-file is working now too. It was an error in the compressor triggered by a non-zero area of more than 7fffh bytes. Fixed by [2ff2b50](https://github.com/SvarDOS/edrdos/commit/2ff2b5079adb911f7a8e3cac5cf09cd166589ecd)

**Comment by ecm-pushbx** [comment]
I adapted the zerocomp packer into an unframed packer that compresses a single, whole file. (Must be compiled on a >16 bit system to support files >= 64 KiB.) At https://hg.pushbx.org/ecm/edrdos/rev/988a9dd9add1

I created an inicomp depacker stage for this file format: https://hg.pushbx.org/ecm/inicomp/rev/ef69f7eef438 This still uses most of the framing of the heatshrink depacker which I based this on. In particular, it will work for both compressed and uncompressed file sizes > 64 KiB, and displays a progress indicator, and checks for error conditions like a buffer overflow.

Here's a bunch of tests running all the interesting methods on the current ecm single-file kernel. As is visible at the start this is built with the kernel-internal compression disabled for both modules. The grep shows how large each inicomp stage is, and doesn't go below 1 KiB even for the simple zerocomp depacker. The .siz file displays performance comparisons (running on dosemu2 on our amd64 Debian server without KVM usable). The edrpackd.siz file lists the final file size of edrpack.sys (the drload variant, as opposed to iniload).

```
edrdos/repo$ cat ovr.bat
@echo off
call c:\autowat.bat
set compressdrbio=0
set compressdrdos=0
edrdos/repo$ grep -F '*****' tmp/*/pedrdos.lst
tmp/apl/pedrdos.lst:  1903          ******************       warning: iniapl: 1328 bytes used for depacker [-w+user]
tmp/bzp/pedrdos.lst:  1903          ******************       warning: inibzp: 1040 bytes used for depacker [-w+user]
tmp/exo/pedrdos.lst:  1903          ******************       warning: iniexo: 1424 bytes used for depacker [-w+user]
tmp/hs/pedrdos.lst:  1903          ******************       warning: inihs: 1344 bytes used for depacker [-w+user]
tmp/lz4/pedrdos.lst:  1903          ******************       warning: inilz4: 1520 bytes used for depacker [-w+user]
tmp/lz/pedrdos.lst:   698          ******************  <1>  warning: localvariables has 14688 bytes [-w+user]
tmp/lz/pedrdos.lst:  1903          ******************       warning: inilz: 3008 bytes used for depacker [-w+user]
tmp/sa2/pedrdos.lst:  1903          ******************       warning: inilzsa2: 1408 bytes used for depacker [-w+user]
tmp/zer/pedrdos.lst:  1903          ******************       warning: inizero: 1072 bytes used for depacker [-w+user]
edrdos/repo$ LC_ALL=C sort tmp/edrpack.spd
    1.07s for 128 runs (    8ms / run), method         zerocomp
    3.64s for 128 runs (   28ms / run), method            lzsa2
    5.80s for 128 runs (   45ms / run), method              lz4
    6.01s for 128 runs (   46ms / run), method              bzp
   18.44s for 128 runs (  144ms / run), method              apl
   20.11s for 128 runs (  157ms / run), method          exodecr
   30.09s for 128 runs (  235ms / run), method       heatshrink
   62.50s for 128 runs (  488ms / run), method              lzd
edrdos/repo$ LC_ALL=C sort tmp/edrpackd.siz
   46128 bytes ( 56.48%), method              lzd
   46688 bytes ( 57.17%), method          exodecr
   47248 bytes ( 57.85%), method              apl
   49040 bytes ( 60.05%), method            lzsa2
   54016 bytes ( 66.14%), method              bzp
   54800 bytes ( 67.10%), method              lz4
   57344 bytes ( 70.21%), method       heatshrink
   61760 bytes ( 75.62%), method         zerocomp
   81664 bytes (100.00%), method             none
edrdos/repo$
```

**Comment by ecm-pushbx** [comment]
Forgot to mention this is the command used to run the tests:

`INICOMP_METHOD="lzd exodecr apl bzp heatshrink lzsa2 lz4 zerocomp" INICOMP_SPEED_TEST=128 ./mak.sh onlypl notracelist`

**Comment by boeckmann** [comment]
> I adapted the zerocomp packer into an unframed packer that compresses a single, whole file. (Must be compiled on a >16 bit system to support files >= 64 KiB.) At https://hg.pushbx.org/ecm/edrdos/rev/988a9dd9add1

The >64K thing is why the _compkern_ does it in two runs, one for the BIO and one for the BDOS, because as a whole they are larger than 64K. Therefore I added a new parameter to the zerocomp function to suppress the terminating zero word.

**Comment by ecm-pushbx** [comment]
> > I adapted the zerocomp packer into an unframed packer that compresses a single, whole file. (Must be compiled on a >16 bit system to support files >= 64 KiB.) At https://hg.pushbx.org/ecm/edrdos/rev/988a9dd9add1
> 
> The >64K thing is why the _compkern_ does it in two runs, one for the BIO and one for the BDOS, because as a whole they are larger than 64K. Therefore I added a new parameter to the zerocomp function to suppress the terminating zero word.

Like I commented I think zerocomp should process its data as a stream rather than in full blocks, ideally. But my scripting requires a non-DOS host regardless so it isn't a great priority for me to support large files in 86 Mode DOS.

```
/* To allow input size >= 64 KiB this tool must be compiled on a 32-bit
   or 64-bit host. Compilation for 86 Mode DOS is possible but may
   truncate the data if it is too large, as size_t is likely a 16-bit
   type and the near buffer in read_file may be limited to < 64 KiB.

   Running for large files on 16-bit DOS would require to change the
   packer to stream data rather than working with entire-file buffers. */
```

**Comment by ecm-pushbx** [comment]
I'm considering to add optimisations for depacked size <= 65_520 Bytes or packed size <= 65_520 Bytes. The build process can "know" the sizes at build time so it shouldn't be a problem to pass these optimisation switches to inicomp.asm and its included depacker.

**Comment by boeckmann** [comment]
> Like I commented I think zerocomp should process its data as a stream rather than in full blocks, ideally.

Yes :)

**Comment by boeckmann** [comment]
> I adapted the zerocomp packer into an unframed packer that compresses a single, whole file. (Must be compiled on a >16 bit system to support files >= 64 KiB.) At https://hg.pushbx.org/ecm/edrdos/rev/988a9dd9add1

Btw. is there a reason you do not add your copyright to these files?

**Comment by ecm-pushbx** [comment]
I don't think I need to do that, if I want to distribute the file under the same license I got it as, right? I will add a copyright or other attribution to sufficiently complex files that I wrote or rewrote my self. But many projects that I contribute to, I don't make such significant edits.

**Comment by boeckmann** [comment]
Like you want, just think its a good thing give people the attribution they deserve. So I for my part have absolutely no problems if you add your name to the files you change.

**Comment by boeckmann** [comment]
Compressed single-file kernel should now be able to boot if copied over a  FreeDOS KERNEL.SYS (but it still gets the boot drive from DL, so a recent SYS with the [FAT32 bootloader adjustments](https://github.com/FDOS/kernel/commit/42d3ae6aacf4dbc948220b813ab29fffdb5d0803) needs to be used.

**Comment by mateuszviste** [comment]
> Compressed single-file kernel should now be able to boot if copied over a FreeDOS KERNEL.SYS (but it still gets the boot drive from DL, so a recent SYS with the [FAT32 bootloader adjustments](https://github.com/FDOS/kernel/commit/42d3ae6aacf4dbc948220b813ab29fffdb5d0803) needs to be used.

So the boot code loads it at seg 60h (like it is with FreeDOS), but the decompression code inflates it at seg 70h?
About the DL thing: is the SYS used by SvarDOS good?
And about the KERNEL.SYS replacement... I am on the fence now: would it be wise to ship the EDR kernel as "KERNEL.SYS"? I think it makes things simpler. But then the next step could be for the package manager to be able to upgrade the kernel (just like it does it already with COMMAND.COM), and that would be very cool, but then if someone wants to switch from one kernel to another, it means that these actions would have to be performed:

"pkg remove kernel"
(critical moment here)
"pkg install edrdos"

Which means that, in between, the system would be effectively "kernel-less"... Besides a power outage at the wrong moment, is there any other risk to do that?

Also, maybe there is a better name for the EDR kernel package, albeit I did not think about it much yet. KERN_EDR? And maybe then rename KERNEL.SVP to KERN_FD? I know these are stupidly petty things compared to actual kernel work, please don't laugh. :)

**Comment by ecm-pushbx** [comment]
> > Compressed single-file kernel should now be able to boot if copied over a FreeDOS KERNEL.SYS (but it still gets the boot drive from DL, so a recent SYS with the [FAT32 bootloader adjustments](https://github.com/FDOS/kernel/commit/42d3ae6aacf4dbc948220b813ab29fffdb5d0803) needs to be used.
> 
> So the boot code loads it at seg 60h (like it is with FreeDOS), but the decompression code inflates it at seg 70h?

Yes, this is based on my input in the comments at https://github.com/SvarDOS/edrdos/issues/72#issuecomment-2248810624 (Albeit I'd rather if the uncompressed kernel were able to do the same.)

**Comment by boeckmann** [comment]
> Yes, this is based on my input in the comments at [#72 (comment)](https://github.com/SvarDOS/edrdos/issues/72#issuecomment-2248810624) (Albeit I'd rather if the uncompressed kernel were able to do the same.)

This can be done in addition. The benefit of doing it while uncompressing is that this saves one relocation in comparison to doing it after the uncompression as a separate step. But we can include a second opportunity to relocate the kernel in the biosinit ICODE segment.

**Comment by boeckmann** [comment]
> So the boot code loads it at seg 60h (like it is with FreeDOS), but the decompression code inflates it at seg 70h?

Correct!

>About the DL thing: is the SYS used by SvarDOS good?

I am not sure. The SYS shipped with the SvarDOS probably not if it is from before february. I could instruct the kernel to take the drive number from BL if it is loaded to 60h. That would mitigate the problem.

> And about the KERNEL.SYS replacement... I am on the fence now: would it be wise to ship the EDR kernel as "KERNEL.SYS"? I think it makes things simpler. But then the next step could be for the package manager to be able to upgrade the kernel (just like it does it already with COMMAND.COM), and that would be very cool, but then if someone wants to switch from one kernel to another, it means that these actions would have to be performed:
> 
> "pkg remove kernel" (critical moment here) "pkg install edrdos"
> 
> Which means that, in between, the system would be effectively "kernel-less"... Besides a power outage at the wrong moment, is there any other risk to do that?

I think it is best to let the kernels install not under the root directory but under the SVARDOS directory. We could provide a tool which lets the user switch between the kernels (and probably command interpreters), lets call it SELECT or something. The tool would look for a KERNEL.SYS under under \SVARDOS\KERNEL\FD and \SVARDOS\KERNEL\EDR (and other subdirectories), then let the user choose between the installed ones, copying the selected to the root dir. Experienced users could of course also do this by themselfes.

> 
> Also, maybe there is a better name for the EDR kernel package, albeit I did not think about it much yet. KERN_EDR? And maybe then rename KERNEL.SVP to KERN_FD? I know these are stupidly petty things compared to actual kernel work, please don't laugh. :)

On the contrary! These are all critical questions on how to not let the users break their systems. All the technical stuff is nothing if it is not manageable by the user. Your naming proposal sounds good to me.

There is also the SYS package which has to be prepared, I saw you assigned it to me. I will do this in the next couple of days.

**Comment by boeckmann** [comment]
@mateuszviste As a test I simply replaced the kernel.sys from http://svardos.org/download/20240201/svardos-20240201-floppy-1.44M-EN_ONLY.zip with a recent single-file build and ran the installer.

This ran through, and the system was bootable. However,  the autoexec.bat and so postinst.bat did not run. I could manually run autoexec.bat and the postinstallation and noted that config.sys was created afterwards. This may be a disagree between the EDR kernel and SvarCOM on when to execute autoexec.bat. I will try a more recent SvarDOS image and have a look at how the kernel calls command.com (which arguments).

As the image was from february, the FAT32 bootsector likely was the old one. So this works.

I will adapt the Github workflow to also generate the single-file kernel...

**Comment by mateuszviste** [comment]
> I think it is best to let the kernels install not under the root directory but under the SVARDOS directory. We could provide a tool which lets the user switch between the kernels (and probably command interpreters), lets call it SELECT or something.

Isn't it a little overkill? After all it's effectively just a single COPY operation now :)
or -- a "pkg remove / pkg install" (as long as the user kept the *.svp packages around). Perhaps a good practice for switching software would be to simply keep the *.svp  packages and install whichever the user needs at the given time... This would require some new feature in pkg, though, so pkg could be instructed to "replace package X using package Y", and do it in sort of atomically.

> However, the autoexec.bat and so postinst.bat did not run.

That's odd. SvarCOM is supposed to execute autoexec.bat whenever it is loaded with /P (that's actually the only thing /P does - instructs SvarCOM to execute C:\AUTOEXEC.BAT). The SvarDOS installer should have created a CONFIG.SYS with a SHELL directive "SHELL=C:\COMMAND.COM /E:512 /P". Maybe the EDR kernel ignores this SHELL directive?

**Comment by mateuszviste** [comment]
This makes me think that having hard-coded C:\AUTOEXEC.BAT in SvarCOM is probably not good, since EDR might just as well assign the D: or E: letter to the booted drive (if the bootable/active partition was the second FAT partition on the disk). There is probably a way to ask the kernel about what drive it booted from, I will have to take a deeper look at this.

**Comment by boeckmann** [comment]
The boot process indeed needs some investigation. I am not sure if the EDR kernel is even able to boot from a drive other than A / C: at the moment. It also does not make use of the partition offset handed over by the boot loader. At least not in the single file version. I will have a look at BDOSLDR.ASM what it does with the partition offset other than loading the DRDOS file.

However, I have extended the Github action to also create a single-file ZIP and disk image. Should be sufficient to start playing with it.

**Comment by boeckmann** [comment]
https://github.com/SvarDOS/edrdos/actions/runs/10100001351

**Comment by boeckmann** [comment]
> That's odd. SvarCOM is supposed to execute autoexec.bat whenever it is loaded with /P (that's actually the only thing /P does - instructs SvarCOM to execute C:\AUTOEXEC.BAT). The SvarDOS installer should have created a CONFIG.SYS with a SHELL directive "SHELL=C:\COMMAND.COM /E:512 /P". Maybe the EDR kernel ignores this SHELL directive?

The CONFIG.SYS gets created eventually, but only after manually running postinst.bat through manually running autoexec.bat once. At least this is the behaviour I encountered. After that, it works fine. EDR spawns the Kernel without /P i think. I can try what happens if I change this to include /P.

**Comment by boeckmann** [comment]
> Isn't it a little overkill? After all it's effectively just a single COPY operation now :) or -- a "pkg remove / pkg install" (as long as the user kept the *.svp packages around). Perhaps a good practice for switching software would be to simply keep the *.svp packages and install whichever the user needs at the given time... This would require some new feature in pkg, though, so pkg could be instructed to "replace package X using package Y", and do it in sort of atomically.

Sounds like a good feature for PKG to implement :) But maybe the user has more to do than switching the kernel, as EDR kernel comes with its own COUNTRY.SYS, for example. I suggest we simply try what works best and start with the least complex solutions :)

**Comment by ecm-pushbx** [comment]
> It also does not make use of the partition offset handed over by the boot loader. At least not in the single file version. I will have a look at BDOSLDR.ASM what it does with the partition offset other than loading the DRDOS file.

I don't think the dual-file version used the partition hidden sectors other than for the BDOS loader. The load unit is likely used only to check for 00h vs 80h (or nonzero?) to select A: or C: as the boot drive.

**Comment by ecm-pushbx** [comment]
> > Yes, this is based on my input in the comments at [#72 (comment)](https://github.com/SvarDOS/edrdos/issues/72#issuecomment-2248810624) (Albeit I'd rather if the uncompressed kernel were able to do the same.)
> 
> This can be done in addition. The benefit of doing it while uncompressing is that this saves one relocation in comparison to doing it after the uncompression as a separate step. But we can include a second opportunity to relocate the kernel in the biosinit ICODE segment.

I don't understand why you need a second opportunity as you put it. Just replace the decompression handling by a relocation that's only run for the uncompressed kernel?

**Comment by boeckmann** [comment]
Sure, replacing via an ifdef would of course be possible.

**Comment by mateuszviste** [comment]
I noticed now that the memory footprint of EDR differs depending on which flavor I use:

two-files EDR kernel: 56K (56848 bytes)
single-file kernel: 59K (59792 bytes)

Hence basically the single-file kernel suffers from a memory footprint increase of over 5% compared to its dual-file counterpart (or at least that's what MEM/C reports).

Is this expected? Can it be somehow mitigated?

**Comment by boeckmann** [comment]
This is unexpected :) Will have to investigate...

**Comment by mateuszviste** [comment]
Nah, the problem appears to be my own stupidity. I retested now using the floppies that you build on github, and there is no difference in the memory footprint between the single-file kernel and the dual-file one. Previously I was testing with two different CONFIG.SYS files, that's why I was getting a difference (the increase was caused by FILES=8 / BUFFERS=10 / LASTDRIVE=Z). Sorry for the noise.

**Comment by boeckmann** [comment]
No problem :) But glad one error less ;)
